Cognite.Simulator.Utils/ModelLibraryBase.cs

```diff
@@ -649,6 +655,67 @@ private async Task<bool> DownloadFileAsync(T modelState, bool isTemporary = fals
             return false;
         }
 
+        /// <summary>
+        /// Processes a model revision by downloading the file, extracting the model information and persisting it.
+        /// If the model revision is already being processed, it will wait for the processing to finish and return the result.
+        /// </summary>
+        /// <param name="modelState">State object representing the file to download</param>
+        /// <param name="token">Cancellation token</param>
+        /// <param name="remote">True if the model revision is remote and should be downloaded from CDF</param>
+        /// <param name="modelRevision">Model revision to process.</param>
+        private async Task<T> ProcessModelRevision(
+            T modelState,
+            CancellationToken token,
+            bool remote = false,
+            SimulatorModelRevision modelRevision = null
+        )
+        {
+            var tcs = _revisionsTasks.GetOrAdd(modelRevision.ExternalId, _ => new TaskCompletionSource<T>());
+            bool shouldProcessModelRevision = false;
+
+            lock (tcs)
+            {
+                if (tcs.Task.Status == TaskStatus.WaitingForActivation)
+                {
+                    shouldProcessModelRevision = true;
+                }
+            }
+
+            if (shouldProcessModelRevision)
+            {
+                try
+                {
+                    var downloaded = await DownloadFileAsync(modelState).ConfigureAwait(false);
+                    if (downloaded)
+                    {
+                        if (remote)
+                        {
+                            UpdateModelParsingInfo(modelState, modelRevision);
+                        }
+                        await ExtractModelInformationAndPersist(modelState, token).ConfigureAwait(false);
+                        UpdateModelParsingInfo(modelState, modelRevision);
+                        _libState.UpdateDestinationRange(
+                            CogniteTime.FromUnixTimeMilliseconds(modelState.UpdatedTime),
+                            CogniteTime.FromUnixTimeMilliseconds(modelState.UpdatedTime));
+                        return modelState;
+                    }
+                }
+                catch (Exception e)
+                {
+                    _logger.LogError("Error processing model revision {ExternalId}: {Message}", modelState.ExternalId, e.Message);
+                }
+
+                return null;
+            }
+
+            if (!shouldProcessModelRevision)
+            {
+                return await tcs.Task.ConfigureAwait(false);
+            }
+
+            return null;
```

Comment:
needs some love. there is an `if (shouldProcessModelRevision)` then there is an `if (!shouldProcessModelRevision)`
1. return null; will never be reached
2. `if (!shouldProcessModelRevision)` seems redundant